hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/MisReplicationHandler.java

```diff
@@ -0,0 +1,186 @@
+              getReplicateCommand(containerInfo, replica));
+      datanodeIdx += 1;
+    }
+    return commandMap;
+
+  }
+  @Override
+  public Map<DatanodeDetails, SCMCommand<?>> processAndCreateCommands(
+      Set<ContainerReplica> replicas, List<ContainerReplicaOp> pendingOps,
+      ContainerHealthResult result, int remainingMaintenanceRedundancy)
+      throws IOException {
+    ContainerInfo container = result.getContainerInfo();
+    ContainerReplicaCount replicaCount = getContainerReplicaCount(container,
+            replicas, pendingOps, remainingMaintenanceRedundancy);
+
+    if (!replicaCount.isSufficientlyReplicated() ||
+            !replicaCount.isOverReplicated()) {
+      LOG.info("Container {} state should be neither under replicated " +
+              "nor over replicated before resolving misreplication." +
+              "Container UnderReplication status: {}," +
+              "Container OverReplication status: {}",
+              container.getContainerID(),
+              replicaCount.isSufficientlyReplicated(),
+              replicaCount.isOverReplicated());
+      return Collections.emptyMap();
+    }
+    final Set<DatanodeDetails> deletionInFlight = new HashSet<>();
+    final Set<DatanodeDetails> addInFlight = new HashSet<>();
+
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        deletionInFlight.add(op.getTarget());
+      } else if (op.getOpType() == ContainerReplicaOp.PendingOpType.ADD) {
+        addInFlight.add(op.getTarget());
+      }
+    }
+
+    List<DatanodeDetails> usedDns = replicas.stream()
+            .map(ContainerReplica::getDatanodeDetails)
+            .collect(Collectors.toList());
+    if (containerPlacement.validateContainerPlacement(usedDns,
+            usedDns.size()).isPolicySatisfied()) {
+      LOG.info("Container {} is currently not misreplicated",
+              container.getContainerID());
+      return Collections.emptyMap();
+    }
+    List<DatanodeDetails> dnsWithPendingOps =
+            Stream.concat(usedDns.stream(), addInFlight.stream())
+            .filter(dn -> !deletionInFlight.contains(dn))
+            .distinct()
+            .collect(Collectors.toList());
+    if (containerPlacement.validateContainerPlacement(dnsWithPendingOps,
+            usedDns.size()).isPolicySatisfied()) {
+      LOG.info("Container {} is currently misreplicated currently" +
+               " but will not be misreplicated after pending ops executed",
+              container.getContainerID());
+      return Collections.emptyMap();
+    }
+    Set<ContainerReplica> sources = filterSources(replicas, deletionInFlight);
+    Set<ContainerReplica> replicasToBeReplicated = containerPlacement
+            .replicasToCopyToFixMisreplication(replicas.stream()
+            .collect(Collectors.toMap(Function.identity(), sources::contains)));
+    usedDns = replicas.stream().filter(r -> !replicasToBeReplicated.contains(r))
```

Comment:
If I recall, usedDNs is used in rackScatter, to setup the racks we have aleady used.

There is also excludedNodes - do we need to pass replicasToBeReplicated in as excluded?

I think it is correct to not pass them as usedDNs, as we are planning to delete them. However we don't want the policy to somehow pick one of those nodes as a new target. We won't be to use it as a target as there is still a replica on it.

Its unlikely the policy would pick one of the replicasToBeReplicated as a target, as the policy would hopefully not try to use the same rack as we are trying to move things away from, but it would be safer, I think, to pass the replicasToBeReplicated list as exluded nodes to be 100% sure.

Reply:
Done